doc/manual/parallel-computing.rst

```diff
@@ -55,31 +42,135 @@ equal the number of CPU cores on the machine.
+---------------------------
+
+Parallel programming in Julia is built on two primitives: *remote
+references* and *remote calls*. A remote reference is used to access
+data stored on a particular process. A remote call is used to call a
+specified function on another (possibly the same) process.
+
+The code below illustrates the basic flavor of this scheme. First,
+julia is launched with 2 worker processes. Then :func:`remotecall`
+is used to construct a 2-by-2 random matrix on process 2. The first
+argument provides the function to be executed, the second argument
+is the process id, and the remaining arguments are passed to the
+executed function. In this case, ``rand(2,2)`` is executed on process
+2.
+
+::
+
+    $ ./julia -p 2
+
     julia> r = remotecall(rand, 2, 2, 2)
     Future(2,1,3,Nullable{Any}())
 
-    julia> s = @spawnat 2 1 .+ fetch(r)
-    Future(2,1,6,Nullable{Any}())
+Notice that :func:`remotecall` returned a ``Future`` object. A ``Future``
+is a type of remote reference (the other type of remote reference is a
+``RemoteChannel``, covered later). The ``Future`` type has the following
+fields.
 
-    julia> fetch(s)
-    2x2 Float64 Array:
-     1.60401  1.50111
-     1.17457  1.15741
-
-The first argument to :func:`remotecall` is the function to call.
-Most parallel programming in Julia does not reference specific processes
-or the number of processes available, but :func:`remotecall` is
-considered a low-level interface providing finer control. The second
-argument to :func:`remotecall` is the index of the process
-that will do the work, and the remaining arguments will be passed
-to the function being called.
-
-As you can see, in the first line we asked process 2 to
-construct a 2-by-2 random matrix, and in the second line we asked it
-to add 1 to it. The result of both calculations is available in the
-two futures, ``r`` and ``s``. The :obj:`@spawnat` macro
-evaluates the expression in the second argument on the process
-specified by the first argument.
+::
+
+    type Future <: AbstractRemoteRef
+        where::Int       # process id where data exists
+        whence::Int      # process id where this ref was instantiated
+        id::Int          # identifier for this Future
+        v::Nullable{Any} # stores the data after fetch() is called
+    end
+
+In our example, ``where = 2`` and ``whence = 1`` because we asked
+process 2 to execute the function, and process 1 (by default) issued
+the request. The field ``v`` is a ``Nullable`` type (see
+:ref:`man-nullable-types`), and it remains empty/null until we
+tell process 2 that we want to retrieve the results of the function
+call. We do this by using :func:`fetch`.
+
+::
+
+    julia> isnull(r.v) # no data fetched yet
+    true
+
+    julia> result = fetch(r)
+    2x2 Array{Float64,2}:
+     0.659427  0.995031
+     0.944037  0.311687
+
+    julia> isnull(r.v) # data has been fetched
+    false
+
+    julia> get(r.v); # if we want to re-access the result
```

Comment:
This comment `# if we want to re-access the result` should be removed. `fetch` on a `Future` anyways checks and returns the cached value. That fact ought to be mentioned. Folks should only use `fetch` and not access the fields of a reference.